flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/hybrid/HybridSourceReader.java

```diff
@@ -60,11 +62,20 @@ public HybridSourceReader(
 
     @Override
     public void start() {
-        setCurrentReader(0);
+        // underlying reader starts on demand with split assignment
     }
 
     @Override
     public InputStatus pollNext(ReaderOutput output) throws Exception {
+        if (currentReader == null) {
+            if (pendingSplits.isEmpty()) {
+                // no underlying reader before split assignment
+                return InputStatus.NOTHING_AVAILABLE;
```

Comment:
Do we want to support a bounded hybrid? Currently, there is no `END_OF_INPUT` return value...

I guess we would need some flag to when we received `noMoreSplits` on last reader index and then return EOI here.

Reply:
The `END_OF_INPUT` from the last reader is passed up. That happens further down, this here is just handling the case where we are initially waiting for splits to arrive.

Reply:
Will be covered by unit test.